integration-tests/src/test/java/com/datastax/oss/driver/api/core/metadata/SerializableMetadataIT.java

```diff
@@ -0,0 +1,153 @@
+import com.datastax.oss.driver.api.core.metadata.schema.TableMetadata;
+import com.datastax.oss.driver.api.core.type.DataTypes;
+import com.datastax.oss.driver.api.testinfra.ccm.CcmRule;
+import com.datastax.oss.driver.api.testinfra.session.SessionRule;
+import com.datastax.oss.driver.categories.ParallelizableTests;
+import com.datastax.oss.driver.internal.core.metadata.schema.DefaultKeyspaceMetadata;
+import com.datastax.oss.driver.internal.core.metadata.schema.DefaultTableMetadata;
+import com.google.common.collect.ImmutableList;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.util.List;
+import java.util.Optional;
+import org.apache.commons.lang3.SerializationUtils;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.rules.RuleChain;
+import org.junit.rules.TestRule;
+
+@Category(ParallelizableTests.class)
+public class SerializableMetadataIT {
+
+  private static final CcmRule ccmRule = CcmRule.getInstance();
+
+  private static final SessionRule<CqlSession> sessionRule = SessionRule.builder(ccmRule).build();
+
+  @ClassRule public static final TestRule chain = RuleChain.outerRule(ccmRule).around(sessionRule);
+
+  @Before
+  public void setupKeyspace() {
+    List<String> statements =
+        ImmutableList.of(
+            "CREATE TABLE IF NOT EXISTS all (val int, doub int, str text, primary key (val,doub)) WITH CLUSTERING ORDER BY (doub DESC)",
+            "CREATE TYPE IF NOT EXISTS numbers_data (doub int, str text)",
+            "CREATE TABLE IF NOT EXISTS smaller (val int primary key, data numbers_data)",
+            "CREATE INDEX IF NOT EXISTS doub_idx on all (doub)",
+            "CREATE OR REPLACE FUNCTION avgState ( state tuple<int,bigint>, val int ) CALLED ON NULL INPUT RETURNS tuple<int,bigint> LANGUAGE java AS"
+                + " 'if (val !=null) { state.setInt(0, state.getInt(0)+1); state.setLong(1, state.getLong(1)+val.intValue()); } return state;'",
+            "CREATE OR REPLACE FUNCTION avgFinal ( state tuple<int,bigint> ) CALLED ON NULL INPUT RETURNS double LANGUAGE java AS"
+                + " 'double r = 0; if (state.getInt(0) == 0) return null; r = state.getLong(1); r/= state.getInt(0); return Double.valueOf(r);'",
+            "CREATE AGGREGATE IF NOT EXISTS average ( int ) SFUNC avgState STYPE tuple<int,bigint> FINALFUNC avgFinal INITCOND (0,0)",
+            "TRUNCATE all",
+            "INSERT INTO all (val,doub,str) VALUES (1,2,'one')",
+            "INSERT INTO all (val,doub,str) VALUES (2,4,'two')",
+            "INSERT INTO all (val,doub,str) VALUES (3,6,'three')",
+            "CREATE MATERIALIZED VIEW IF NOT EXISTS big AS SELECT * FROM all"
+                + " WHERE val is not null and doub is not null and val > 2"
+                + " PRIMARY KEY (val,doub)");
```

Comment:
Schema-related testing is always annoying because it has to work against all the server versions in our CI suite. In particular, the functions, aggregates and views used here are going to be a problem with legacy Cassandra versions.

I don't like to piggyback on existing tests, but for this particular case we could avoid a lot of headaches by reusing `DescribeIT`, which already has the logic to deal with those issues. It could be as simple as throwing a serialization roundtrip in the middle of the existing method, or we could add a separate one.

As a bonus, comparing the `describeWithChildren` output of the top-level keyspace gives us a pretty reliable deep comparison, without having to enumerate all the children.

Reply:
Hmmm, interesting... I'll look more at DescribeIT and see what's involved in such a migration.  Doesn't sound too bad on the face of it.